source/opt/dominator_tree.cpp

```diff
@@ -0,0 +1,383 @@
+template <typename BBType, typename SuccessorLambda, typename PreLambda,
+          typename PostLambda>
+static void DepthFirstSearch(const BBType* bb, SuccessorLambda successors,
+                             PreLambda pre, PostLambda post) {
+  // Ignore backedge operation.
+  auto nop_backedge = [](const BBType*, const BBType*) {};
+  CFA<BBType>::DepthFirstTraversal(bb, successors, pre, post, nop_backedge);
+}
+
+// Wrapper around CFA::DepthFirstTraversal to provide an interface to perform
+// depth first search on generic BasicBlock types. This overload is for only
+// performing user defined post order.
+//
+// BBType - BasicBlock type. Will either be ir::BasicBlock or DominatorTreeNode
+// SuccessorLambda - Lamdba matching the signature of 'const
+// std::vector<BBType>*(const BBType *A)'. Will return a vector of the nodes
+// succeding BasicBlock A.
+// PostLambda - Lamdba matching the signature of 'void (const BBType*)' will be
+// called on each node traversed after their children.
+template <typename BBType, typename SuccessorLambda, typename PostLambda>
+static void DepthFirstSearchPostOrder(const BBType* bb,
+                                      SuccessorLambda successors,
+                                      PostLambda post) {
+  // Ignore preorder operation.
+  auto nop_preorder = [](const BBType*) {};
+  DepthFirstSearch(bb, successors, nop_preorder, post);
+}
+
+// Small type trait to get the function class type.
+template <typename BBType>
+struct GetFunctionClass {
+  using FunctionType = ir::Function;
+};
+
+// This helper class is basically a massive workaround for the current way that
+// depth first is implemented.
+template <typename BBType>
+class BasicBlockSuccessorHelper {
+  // This should eventually become const ir::BasicBlock.
+  using BasicBlock = BBType;
+  using Function = typename GetFunctionClass<BBType>::FunctionType;
+
+  using BasicBlockListTy = std::vector<BasicBlock*>;
+  using BasicBlockMapTy = std::map<const BasicBlock*, BasicBlockListTy>;
+
+ public:
+  BasicBlockSuccessorHelper(Function& func, BasicBlock* dummy_start_node,
+                            bool post);
+
+  // CFA::CalculateDominators requires std::vector<BasicBlock*>.
+  using GetBlocksFunction =
+      std::function<const std::vector<BasicBlock*>*(const BasicBlock*)>;
+
+  // Returns the list of predecessor functions.
+  GetBlocksFunction GetPredFunctor() {
+    return [&](const BasicBlock* bb) {
+      BasicBlockListTy* v = &predecessors_[bb];
+      return v;
+    };
+  }
+
+  // Returns a vector of the list of successor nodes from a given node.
+  GetBlocksFunction GetSuccessorFunctor() {
+    return [&](const BasicBlock* bb) {
+      BasicBlockListTy* v = &successors_[bb];
+      return v;
+    };
+  }
+
+ private:
+  bool invert_graph_;
+  BasicBlockMapTy successors_;
+  BasicBlockMapTy predecessors_;
+
+  // Build a bi-directional graph from the CFG of F.
+  // If invert_graph_ is true, all edge are reverted (successors becomes
```

Comment:
Is it really bi-directional? If edges get reversed in one case it is just a DAG.

reverted -> reversed.

Reply:
The bi-directional graph refers to the successor and predecessor maps in that helper class. They'll always maintain a link forward and back